File: .claude/skills/ios-admob-ads-skill/templates/AdsCore/AdsConfig.swift
import Foundation

/// Remote-tunable ad configuration (`ads_config` JSON).
///
/// Decoding is lenient on purpose: any missing or mistyped key falls back to the safe default
/// for that key only, so a half-edited Remote Config block can never make ads more aggressive
/// or break unrelated keys. `AdsConfig.safeDefault` is what a build that never reached Remote
/// Config runs with: ads off.
public struct AdsConfig: Sendable, Equatable, Codable {
    public var enabled: Bool
    /// No full-screen ad of any format within this many seconds of the last full-screen dismissal.
    public var fullScreenCooldownSeconds: Double
    public var appOpen: AppOpen
    public var interstitial: Interstitial
    public var native: Native
    public var banner: Banner
    public var rewarded: Rewarded
    /// Unit ID overrides keyed by `AdFormat.rawValue`. Unknown keys are kept, not rejected.
    public var unitIDs: [String: String]

    public static let safeDefault = AdsConfig(
        enabled: false,
        fullScreenCooldownSeconds: 30,
        appOpen: .init(),
        interstitial: .init(),
        native: .init(),
        banner: .init(),
        rewarded: .init(),
        unitIDs: [:]
    )

    public init(
        enabled: Bool,
        fullScreenCooldownSeconds: Double,
        appOpen: AppOpen,
        interstitial: Interstitial,
        native: Native,
        banner: Banner,
        rewarded: Rewarded,
        unitIDs: [String: String]
    ) {
        self.enabled = enabled
        self.fullScreenCooldownSeconds = fullScreenCooldownSeconds
        self.appOpen = appOpen
        self.interstitial = interstitial
        self.native = native
        self.banner = banner
        self.rewarded = rewarded
        self.unitIDs = unitIDs
    }

    /// Decodes a Remote Config JSON string; any failure yields `safeDefault`.
    public static func decode(_ json: String?) -> AdsConfig {
        guard let data = json?.data(using: .utf8), !data.isEmpty,
              let config = try? JSONDecoder().decode(AdsConfig.self, from: data)
        else { return .safeDefault }
        return config
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let d = AdsConfig.safeDefault
        enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
        fullScreenCooldownSeconds = c.lenient(Double.self, .fullScreenCooldownSeconds) ?? d.fullScreenCooldownSeconds
        appOpen = c.lenient(AppOpen.self, .appOpen) ?? d.appOpen
        interstitial = c.lenient(Interstitial.self, .interstitial) ?? d.interstitial
        native = c.lenient(Native.self, .native) ?? d.native
        banner = c.lenient(Banner.self, .banner) ?? d.banner
        rewarded = c.lenient(Rewarded.self, .rewarded) ?? d.rewarded
        unitIDs = c.lenient([String: String].self, .unitIDs) ?? d.unitIDs
    }

    // MARK: - Formats

    public struct AppOpen: Sendable, Equatable, Codable {
        public var enabled = true
        public var minimumBackgroundSeconds: Double = 45
        public var minimumSecondsBetween: Double = 300
        public var maxPerDay = 4
        public var skipFirstSessions = 3

        public init() {}

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            let d = Self()
            enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
            minimumBackgroundSeconds = c.lenient(Double.self, .minimumBackgroundSeconds) ?? d.minimumBackgroundSeconds
            minimumSecondsBetween = c.lenient(Double.self, .minimumSecondsBetween) ?? d.minimumSecondsBetween
            maxPerDay = c.lenient(Int.self, .maxPerDay) ?? d.maxPerDay
            skipFirstSessions = c.lenient(Int.self, .skipFirstSessions) ?? d.skipFirstSessions
        }
    }

    public struct Interstitial: Sendable, Equatable, Codable {
        public var enabled = true
        public var minimumSecondsBetween: Double = 180
        /// AdMob allows at most one interstitial per two user actions. Clamped to >= 2 by `AdGate`.
        public var minimumCompletionsBetween = 2
        public var maxPerDay = 6
        public var minimumSessionSeconds: Double = 60
        public var minimumLifetimeCompletions = 3

        public init() {}

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            let d = Self()
            enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
            minimumSecondsBetween = c.lenient(Double.self, .minimumSecondsBetween) ?? d.minimumSecondsBetween
            minimumCompletionsBetween = c.lenient(Int.self, .minimumCompletionsBetween) ?? d.minimumCompletionsBetween
            maxPerDay = c.lenient(Int.self, .maxPerDay) ?? d.maxPerDay
            minimumSessionSeconds = c.lenient(Double.self, .minimumSessionSeconds) ?? d.minimumSessionSeconds
            minimumLifetimeCompletions = c.lenient(Int.self, .minimumLifetimeCompletions) ?? d.minimumLifetimeCompletions
        }
    }

    public struct Native: Sendable, Equatable, Codable {
        public var enabled = true
        /// Placement names allowed to render. Names, not indexes, so one can be switched off remotely.
        public var placements: [String] = []
        /// 1-based display row of the first ad (6 = after five content rows).
        public var firstRow = 6
        public var everyRows = 10
        public var maxPerScreen = 3

        public init() {}

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            let d = Self()
            enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
            placements = c.lenient([String].self, .placements) ?? d.placements
            firstRow = c.lenient(Int.self, .firstRow) ?? d.firstRow
            everyRows = c.lenient(Int.self, .everyRows) ?? d.everyRows
            maxPerScreen = c.lenient(Int.self, .maxPerScreen) ?? d.maxPerScreen
        }
    }

    public struct Banner: Sendable, Equatable, Codable {
        public var enabled = false
        public var placements: [String] = []

        public init() {}

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            let d = Self()
            enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
            placements = c.lenient([String].self, .placements) ?? d.placements
        }
    }

    public struct Rewarded: Sendable, Equatable, Codable {
        public var enabled = true
        public var maxPerDay = 3
        /// For UI copy only. The server decides and clamps the real grant.
        public var grantAmount = 5

        public init() {}

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            let d = Self()
            enabled = c.lenient(Bool.self, .enabled) ?? d.enabled
            maxPerDay = c.lenient(Int.self, .maxPerDay) ?? d.maxPerDay
            grantAmount = c.lenient(Int.self, .grantAmount) ?? d.grantAmount
        }
    }
}

extension KeyedDecodingContainer {
    /// Returns nil for a missing key *or* a value of the wrong type, instead of throwing.
    fileprivate func lenient<T: Decodable>(_ type: T.Type, _ key: Key) -> T? {
        (try? decodeIfPresent(type, forKey: key)) ?? nil
    }
}

File: App/oneai_v2/lib/features/ads/ads_config.dart
// Mirrors Backend/oneai_backend/remote-config/ads_config.defaults.json.
// One Remote Config JSON parameter; a domain that fails to parse keeps its
// own defaults. Defaults are the SAFE side: ads off, wide spacing, low caps.
import 'dart:convert';

import 'package:one_ai/data/firebase/json_read.dart';

enum AdFormat { appOpen, interstitial, rewarded, banner, native }

class AppOpenRules {
  const AppOpenRules({
    this.enabled = true,
    this.minimumBackgroundSeconds = 45,
    this.minimumSecondsBetween = 300,
    this.maxPerDay = 4,
    this.skipFirstSessions = 3,
  });
  factory AppOpenRules.fromJson(Map<String, dynamic> j) => AppOpenRules(
        enabled: readBool(j, 'enabled', orElse: true),
        minimumBackgroundSeconds: readInt(j, 'minimumBackgroundSeconds') ?? 45,
        minimumSecondsBetween: readInt(j, 'minimumSecondsBetween') ?? 300,
        maxPerDay: readInt(j, 'maxPerDay') ?? 4,
        skipFirstSessions: readInt(j, 'skipFirstSessions') ?? 3,
      );
  final bool enabled;
  final int minimumBackgroundSeconds;
  final int minimumSecondsBetween;
  final int maxPerDay;
  final int skipFirstSessions;
}

class InterstitialRules {
  const InterstitialRules({
    this.enabled = true,
    this.minimumSecondsBetween = 180,
    this.minimumCompletionsBetween = 2,
    this.minimumLifetimeCompletions = 3,
    this.minimumSessionSeconds = 30,
    this.maxPerDay = 8,
    this.placements = const ['summaryEnter', 'summaryExit', 'afterShare', 'settingsExit'],
  });
  factory InterstitialRules.fromJson(Map<String, dynamic> j) => InterstitialRules(
        enabled: readBool(j, 'enabled', orElse: true),
        minimumSecondsBetween: readInt(j, 'minimumSecondsBetween') ?? 180,
        minimumCompletionsBetween: readInt(j, 'minimumCompletionsBetween') ?? 2,
        minimumLifetimeCompletions: readInt(j, 'minimumLifetimeCompletions') ?? 3,
        minimumSessionSeconds: readInt(j, 'minimumSessionSeconds') ?? 30,
        maxPerDay: readInt(j, 'maxPerDay') ?? 8,
        placements: j.containsKey('placements') ? readStringList(j, 'placements') : const ['summaryEnter', 'summaryExit', 'afterShare', 'settingsExit'],
      );
  final bool enabled;
  final int minimumSecondsBetween;
  final int minimumCompletionsBetween;
  final int minimumLifetimeCompletions;
  final int minimumSessionSeconds;
  final int maxPerDay;
  final List<String> placements;
}

class RewardedRules {
  const RewardedRules({this.enabled = true, this.maxPerDay = 5});
  factory RewardedRules.fromJson(Map<String, dynamic> j) =>
      RewardedRules(enabled: readBool(j, 'enabled', orElse: true), maxPerDay: readInt(j, 'maxPerDay') ?? 5);
  final bool enabled;
  final int maxPerDay;
}

class BannerRules {
  const BannerRules({this.enabled = true, this.refreshSeconds = 60, this.placements = const ['summaryTab']});
  factory BannerRules.fromJson(Map<String, dynamic> j) => BannerRules(
        enabled: readBool(j, 'enabled', orElse: true),
        refreshSeconds: readInt(j, 'refreshSeconds') ?? 60,
        placements: j.containsKey('placements') ? readStringList(j, 'placements') : const ['summaryTab'],
      );
  final bool enabled;
  final int refreshSeconds;
  final List<String> placements;
}

class NativeRules {
  const NativeRules({this.enabled = false, this.firstRow = 6, this.everyRows = 10, this.maxPerScreen = 3, this.placements = const ['minutesList']});
  factory NativeRules.fromJson(Map<String, dynamic> j) => NativeRules(
        enabled: readBool(j, 'enabled'),
        firstRow: readInt(j, 'firstRow') ?? 6,
        everyRows: readInt(j, 'everyRows') ?? 10,
        maxPerScreen: readInt(j, 'maxPerScreen') ?? 3,
        placements: j.containsKey('placements') ? readStringList(j, 'placements') : const ['minutesList'],
      );
  final bool enabled;
  final int firstRow;
  final int everyRows;
  final int maxPerScreen;
  final List<String> placements;
}

class AdsConfig {
  const AdsConfig({
    this.enabled = false,
    this.fullScreenCooldownSeconds = 30,
    this.appOpen = const AppOpenRules(),
    this.interstitial = const InterstitialRules(),
    this.rewarded = const RewardedRules(),
    this.banner = const BannerRules(),
    this.native = const NativeRules(),
  });

  /// Parse the Remote Config string. Any domain that fails keeps its defaults;
  /// an unparseable string yields the full defaults (ads OFF).
  factory AdsConfig.parse(String? raw) {
    if (raw == null || raw.trim().isEmpty) return const AdsConfig();
    Map<String, dynamic> j;
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map) return const AdsConfig();
      j = Map<String, dynamic>.from(decoded);
    } catch (_) {
      return const AdsConfig();
    }
    T domain<T>(String key, T Function(Map<String, dynamic>) parse, T fallback) {
      final d = readObject(j, key);
      if (d == null) return fallback;
      try {
        return parse(d);
      } catch (_) {
        return fallback;
      }
    }
    return AdsConfig(
      enabled: readBool(j, 'enabled'),
      fullScreenCooldownSeconds: readInt(j, 'fullScreenCooldownSeconds') ?? 30,
      appOpen: domain('appOpen', AppOpenRules.fromJson, const AppOpenRules()),
      interstitial: domain('interstitial', InterstitialRules.fromJson, const InterstitialRules()),
      rewarded: domain('rewarded', RewardedRules.fromJson, const RewardedRules()),
      banner: domain('banner', BannerRules.fromJson, const BannerRules()),
      native: domain('native', NativeRules.fromJson, const NativeRules()),
    );
  }

  final bool enabled;
  final int fullScreenCooldownSeconds;
  final AppOpenRules appOpen;
  final InterstitialRules interstitial;
  final RewardedRules rewarded;
  final BannerRules banner;
  final NativeRules native;
}
